### DISCUSSION TITLE
Scaling as a window rule

### DISCUSSION THREAD

**Discussion by tversteeg** [YaLTeR/niri#276] [Ideas]
It would be nice to set scaling, similar to output scaling, on a per-window basis using window rules.

For example, normally I like the scaling of 1, but some applications use tiny icons and I would like to scale them specifically to 2.

Is this allowed in the Wayland spec?

**Comment by YaLTeR** [comment]
This is kind of tricky. The newest wl-compositor version allows to do this because the compositor explicitly sends the preferred scale for each surface, however barely any apps use that yet.

Before, surfaces were supposed to determine their scale based on which outputs they are "entered" on. Messing with real outputs is not the best idea because more things depend on them, but is possible. Unfortunately, some clients also do something completely wrong like take a max of all advertised outputs' scales.

And I guess you're also looking for not only getting the clients to render at 2x scale, but also drawing them with 2x zoom, since by default 2x scaled clients on 1x outputs would just be downscaled by 2. This one may require annoying refactors because now the logical coordinates, sizes and bounds sent to windows also need to be scaled (if a window pretends to be 2x larger then it has 2x less space on the monitor).

So to sum up, it appears to be a more complex problem than it initially seems. I'd probably prefer only doing it for new clients to avoid messing with the outputs too much.

**Reply by MangoTCF** [reply]
Maybe we can have a command for changing the next launched window's scale? Like, `niri msg action next-window-scale <SCALE>`

**Reply by YaLTeR** [reply]
Well the problems are all the same as I described.
